### PR TITLE
Add a new option to improve type argument inference debugging.

### PR DESCRIPTION
**Pull request by wmdietlGC**

### CODE REVIEW

**Comment by wmdietl** [comment]
https://github.com/typetools/checker-framework-inference/pull/48 will fix checker-framework-inference.
I think this is ready to merge.
Do not squash changes to preserve descriptions.

**Comment by smillst** [comment]
This pull request is three distinct changes.  I cherry-picked one of the changes and pushed it to master (35f94bd8cc43fc7af5985f19a0be467069eb0ed8).  I rebased the type argument inference fix into one commit 1c0dac4db1a0032af0c814e09372cdae80f2f5ad.  Those changes look good to me, so once those test pass (model the expected downstream failure), I'll merge that commit to master.  Once that is down, I'll review the remaining changes.

File: framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java [hunk 1]
```diff
@@ -113,15 +120,33 @@
         final AnnotatedTypeMirror assignedTo =
                 TypeArgInferenceUtil.assignedTo(typeFactory, typeFactory.getPath(expressionTree));
 
+        if (showInferenceSteps) {
+            System.err.println("DTAI: expression: " + expressionTree.toString().replace("\n", " "));
```
Comment (smillst) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Does using System.err have benefits over using a diagnostic note?

Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/typeinference/DefaultTypeArgumentInference.java]: Changed it to use the javac messager.
This revealed the change in https://github.com/typetools/checker-framework/pull/1289 , which is included here to make this code compile.

**Pull request merged**